### PR TITLE
Insert classlist from csv

### PR DESCRIPTION
**Pull request by mamerisawesome**

### CODE REVIEW

**Comment by mamerisawesome** [comment]
@jgjadaoag 

Kindly check this pull request. Thanks.

File: controllers/class.js [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
+    destination: (req, file, cb) => {
+		let destFolder =__dirname + '/../uploads/csv';
+		if (!fs.existsSync(destFolder)) {
+			fs.mkdirSync(destFolder);
+		}
+
+		cb(null, destFolder);
+    },
+    filename: (req, file, cb) => {
+		cb(null,file.originalname);
```
Comment (ghost) [inline controllers/class.js]: What about file conflicts?

File: controllers/class.js [hunk 2]
```diff
@@ -146,57 +162,105 @@ exports.write_to_csv = (req, res, next) => {
...
-	
-	function send_response (err, result, args, last_query) {
+	function initialize_shell (err) {
+		if(err) {
+			winston.error('Error uploading file.');
+			return next(err);
+		}
+
+		path = req.file.path;
+		sh.exec('cd ./', generate_container);
```
Comment (ghost) [inline controllers/class.js]: Para saan to? Bakit ka mag cd sa current directory?

File: controllers/class.js [hunk 3]
```diff
@@ -146,57 +162,105 @@ exports.write_to_csv = (req, res, next) => {
...
+		sh.exec('touch uploads/csv', overwrite_permission);
+	}
+
+	function overwrite_permission (err) {
+		if (err) {
+			winston.error('Error in generating csv folder');
+			return next(err);
+		}
+
+		sh.exec('sudo chmod 755 helpers/classlist.js', generate_query);
```
Comment (ghost) [inline controllers/class.js]: Just put this step into setup or at the start of server.js . Nauulit siya kahit di naman kailangan

File: controllers/class.js [hunk 4]
```diff
@@ -146,57 +162,105 @@ exports.write_to_csv = (req, res, next) => {
...
+	}
+
+	function clean_sql (err) {
+		if (err) {
+			winston.error('Error in inserting classlist from CSV');
+			sh.rm('database/classlist.sql', path);
+			return next(err);
+        }
+
+		sh.exec('rm database/classlist.sql', clean_csv);
```
Comment (ghost) [inline controllers/class.js]: Pano kung may isa pang nag-execute ng generate_query tapos nag-execute yung clean_sql?

File: controllers/class.js [hunk 5]
```diff
@@ -146,57 +162,105 @@ exports.write_to_csv = (req, res, next) => {
...
+		sh.exec('sudo chmod 755 helpers/classlist.js', generate_query);
+	}
+
+	function generate_query (err) {
+		if (err) {
+			winston.error('Error in setting permissions of helper function');
+			return next(err);
+		}
+
+		sh.exec('node helpers/classlist.js ' + path + ' > database/classlist.sql', execute_query);
```
Comment (ghost) [inline controllers/class.js]: Pano kung may isa na nag-execute ng generate_query tapos may isa pang nag-execute ng generate_query?

File: controllers/class.js (hunk 1 above)
Comment (mamerisawesome) [inline controllers/class.js]: fs.existsSync(destFolder) checks if directory exists. So hindi siya gagawa ng bagong folder kung nag-eexist na.

File: controllers/class.js (hunk 2 above)
Comment (mamerisawesome) [inline controllers/class.js]: For safety measures. Pero I realized na it's not really needed so I'll remove it.

File: controllers/class.js (hunk 3 above)
Comment (mamerisawesome) [inline controllers/class.js]: Noted.

**Comment by mamerisawesome** [comment]
@jgjadaoag Kindly check.

File: controllers/class.js [hunk 6]
```diff
@@ -260,38 +302,78 @@ exports.create_class = (req, res, next) => {
...
-	
-	function send_response (err, result, args, last_query) {
+	function overwrite_permission (err) {
+		if (err) {
+			winston.error('Error in uploading csv file');
+			return next(err);
+		}
+
+		path = req.file.path;
+		sh.exec('sudo chmod 755 helpers/classlist.js', generate_query);
```
Comment (jgjadaoag) [inline controllers/class.js]: No need to run multiple times

File: controllers/class.js [hunk 7]
```diff
@@ -260,38 +302,78 @@ exports.create_class = (req, res, next) => {
...
+	}
+
+	function generate_query (err) {
+		if (err) {
+			winston.error('Error in setting permissions of helper function');
+			return next(err);
+		}
+
+		sh.exec('node helpers/classlist.js ' + path + ' > database/classlist-' + timestamp + '.sql', filter_query);
```
Comment (jgjadaoag) [inline controllers/class.js]: Meron pa ring possibility na magkaparehas ang file name dahil nagkasabay

File: controllers/class.js [hunk 8]
```diff
@@ -260,38 +302,78 @@ exports.create_class = (req, res, next) => {
...
+	}
+
+    function filter_query (err) {
         if (err) {
-            winston.error('Error in inserting classlist from CSV');
+			winston.error('Error in parsing CSV file and converting it to into SQL format');
+			return next(err);
+		}
+
+		sh.exec('sudo mysql -uroot < database/classlist-' + timestamp + '.sql', execute_query);
```
Comment (jgjadaoag) [inline controllers/class.js]: use config for username and password for easy changing

File: controllers/class.js [hunk 9]
```diff
@@ -156,39 +198,24 @@ exports.write_to_csv = (req, res, next) => {
...
-	sh.exec('mkdir -p uploads/csv/', query);
-
-	}
-
-	function query(){
+	function start(){
 		mysql.use('master')
 			.query(
 				'SELECT * FROM class ORDER BY class_id',
 				[],
```
Comment (mamerisawesome) [inline controllers/class.js]: kung walang ipapasa, pwedeng callback na kaagad

File: controllers/class.js [hunk 10]
```diff
@@ -305,37 +332,68 @@ exports.create_class = (req, res, next) => {
...
+		sh.exec('sudo mysql -uroot < database/classlist-' + timestamp + '.sql', execute_query);
+	}
+
+	function execute_query (err) {
+		if (err) {
+			winston.error('Error in filtering query');
+			return next(err);
+		}
+
+		sh.exec('sudo mysql -uroot < database/classlist-' + timestamp + '.sql', clean_sql);
```
Comment (mamerisawesome) [inline controllers/class.js:388]: dumoble

File: controllers/class.js [hunk 11]
```diff
@@ -1,11 +1,34 @@
...
+		let destFolder =__dirname + '/../uploads/csv';
+		if (!fs.existsSync(destFolder)) {
+			fs.mkdirSync(destFolder);
+		}
+
+		cb(null, destFolder);
+    },
+    filename: (req, file, cb) => {
+        file.originalname = 'classlist-' + date.get_today();
+        if(sh.exec('test -f ' + 'uploads/csv/classlist-' + date.get_today()).code == 0) {
```
Comment (jgjadaoag) [inline controllers/class.js]: use ===

File: controllers/class.js [hunk 12]
```diff
@@ -98,31 +147,30 @@ exports.update_class = (req, res, next) => {
...
+	function send_response (err, result, args, last_query) {
+		if (err) {
+			winston.error('Error in updating class', last_query);
+			return next(err);
 		}
 
-start();
+		if (result.affectedRows === 0) {
+			return res.status(404)
+                .error({code: 'CLASS404', message: 'Class not found'})
```
Comment (jgjadaoag) [inline controllers/class.js]: be consistent, use only spaces or use only tabs

File: controllers/class.js [hunk 13]
```diff
@@ -182,39 +230,23 @@ exports.write_to_csv = (req, res, next) => {
...
 		var i=0;
-		
+
 		if(err){
 			winston.error('Selection query of classes failed', last_query);
 			return next(err);
-		}		
+		}
 
 		result.forEach(function (element) {
```
Comment (jgjadaoag) [inline controllers/class.js]: use the shorthand anonymous function declaration

**Pull request merged**